### ISSUE TITLE
mpirun outputs petsc errors

### ISSUE THREAD

**Issue by fanronghong** [mfem/mfem#1397]
Hi all,

I have the below simple example code (see the 1st code block). 
When I run in serial, the code can give the right result but also give an error message below (see the 2nd code block).
If I run in parallel, it won't give a right output (only give error messages below, see the 3rd code block).

I don't know where is wrong and how to fix it.
I think it's a very easy question, but I am not familar with MPI. Thanks in advance.
```c++
#include "mfem.hpp"
#include <iostream>
using namespace std;
using namespace mfem;

class UserDefined_KSP_Monitor: public PetscSolverMonitor
{
private:
    ParBilinearForm* blf;
    ParLinearForm* lf;
    ParMesh* mesh;
    ParFiniteElementSpace* fsp;

    int num_procs, myid;

public:
    UserDefined_KSP_Monitor(ParBilinearForm *a, ParLinearForm *b)
            : PetscSolverMonitor(true,true), blf(a), lf(b)
    {
        fsp = blf->ParFESpace();
        mesh = fsp->GetParMesh();
        MPI_Comm_size(fsp->GetComm(),&num_procs);
        MPI_Comm_rank(mesh->GetComm(),&myid);
    }
    virtual ~UserDefined_KSP_Monitor() {}

    virtual void MonitorSolution(PetscInt it, PetscReal norm, const Vector& x)
    {
        if (myid == 0)
        {
            cout << setiosflags(ios::right) << setw(4) << it << "-th iteration: ";
            cout << " MonitorSolution(),";
            ostream::fmtflags old = cout.flags();
            cout << "l2 norm of x_n: "
                 << resetiosflags(ios::right) << setiosflags(ios::scientific|ios::left) << setw(12) << setfill(' ') << x.Norml2(); 
            cout.flags(old);
        }
    }

    virtual void MonitorResidual(PetscInt it, PetscReal norm, const Vector& res)
    {
        if (myid == 0)
        {
            ostream::fmtflags old = cout.flags();
            cout << ";    MonitorResidual(),l2 norm of residual: "
                 << setiosflags(ios::scientific|ios::left) << setw(12) << setfill(' ') << res.Norml2();
            cout.flags(old);
        }
    }

    virtual void MonitorSolver(PetscSolver* solver)
    {
        if (myid == 0) cout << ";    MonitorSolver()" << endl;
    }
};

class UserDefined_Preconditioner: public PetscLinearSolver
{
private:
    PetscParMatrix& A1;
    PetscLinearSolver* ksp1;

public:
    UserDefined_Preconditioner(MPI_Comm comm, PetscParMatrix& A1_, const std::string &prefix)
    : PetscLinearSolver(comm, prefix, false), A1(A1_)
    {
        ksp1 = new PetscLinearSolver(A1, prefix);
    }
    ~UserDefined_Preconditioner() { delete ksp1; }

    virtual void SetOperator(const Operator& op) {}

    virtual void Mult(const Vector& b, Vector& x) const
    {
        cout << "\n---> UserDefined_Preconditioner::Mult()" << endl;
        ksp1->Mult(b, x);
    }
};

int main(int argc, char** argv)
{
    int num_procs, myid;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);

    const char* options_src = "./petsc_example_opts";
    MFEMInitializePetsc(NULL, NULL, options_src, NULL);

    Mesh mesh(100, 100, Element::TRIANGLE, true, 1.0, 1.0);
    int dim = mesh.Dimension();
    ParMesh pmesh(MPI_COMM_WORLD, mesh);

    H1_FECollection h1_fec(1, dim);
    ParFiniteElementSpace h1_space(&pmesh, &h1_fec);

    Array<int> ess_tdof_list;
    {
        Array<int> ess_bdr(pmesh.bdr_attributes.Max());
        ess_bdr = 1;
        h1_space.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
    }

    ParGridFunction u(&h1_space);
    u = 0.0;

    ParLinearForm lf(&h1_space);
    ConstantCoefficient one(1.0);
    lf.AddDomainIntegrator(new DomainLFIntegrator(one));
    lf.Assemble();

    ParBilinearForm blf(&h1_space);
    ConstantCoefficient diff(1E+2);
    Vector adv_(dim);  adv_ = 1.0;
    VectorConstantCoefficient adv(adv_);
    blf.AddDomainIntegrator(new DiffusionIntegrator(diff));
    blf.AddDomainIntegrator(new ConvectionIntegrator(adv));
    blf.Assemble();

    PetscParVector b(&h1_space), x(&h1_space);
    PetscParMatrix A;
    blf.SetOperatorType(Operator::PETSC_MATAIJ);
    blf.FormLinearSystem(ess_tdof_list, u, lf, A, x, b);

    ParBilinearForm spd_blf(&h1_space);
    spd_blf.AddDomainIntegrator(new DiffusionIntegrator(diff));
    spd_blf.Assemble();

    PetscParMatrix B;
    spd_blf.SetOperatorType(Operator::PETSC_MATAIJ);
    spd_blf.FormSystemMatrix(ess_tdof_list, B);
    PetscLinearSolver *pc = new UserDefined_Preconditioner(MPI_COMM_WORLD, B, "userpc_");

    PetscLinearSolver ksp(MPI_COMM_WORLD, "main_");
    ksp.SetOperator(A);
    ksp.SetPreconditioner(*pc);

    UserDefined_KSP_Monitor user_monitor(&blf, &lf);
    ksp.SetMonitor(&user_monitor);

    ksp.Mult(b, x);
    if (ksp.GetConverged() == 1 && myid == 0) cout << "Successfully Converged." << endl;

    delete pc;

    MFEMFinalizePetsc();
    MPI_Finalize();
    return 0;
}
```
```*** The MPI_Comm_rank() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[yoga:11666] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/linearsolvertable.html for possible LU and Cholesky solvers
[0]PETSC ERROR: Could not locate a solver package for factorization type ILU and matrix type mpiaij.
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.12.3, unknown 
[0]PETSC ERROR: Unknown Name on a arch-linux2-c-debug named yoga by fan Sat Apr  4 09:19:39 2020
[0]PETSC ERROR: Configure options --download-fblaslapack=1
[0]PETSC ERROR: #1 MatGetFactor() line 4416 in /home/fan/softwares/petsc/src/mat/interface/matrix.c
[0]PETSC ERROR: #2 PCSetUp_ILU() line 133 in /home/fan/softwares/petsc/src/ksp/pc/impls/factor/ilu/ilu.c
[0]PETSC ERROR: #3 PCSetUp() line 894 in /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[0]PETSC ERROR: #4 KSPSetUp() line 376 in /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[0]PETSC ERROR: #5 KSPSolve() line 703 in /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[0]PETSC ERROR: #6 virtual void mfem::PetscLinearSolver::Mult(const mfem::Vector&, mfem::Vector&) const() line 2487 in /home/fan/Nutstore Files/github_nutstore/mfem_source_code/linalg/petsc.cpp


MFEM abort: Error in PETSc. See stacktrace above.
 ... in function: virtual void mfem::PetscLinearSolver::Mult(const mfem::Vector&, mfem::Vector&) const
 ... in file: /home/fan/Nutstore Files/github_nutstore/mfem_source_code/linalg/petsc.cpp:2487

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
[1]PETSC ERROR: ------------------------------------------------------------------------
[1]PETSC ERROR: Caught signal number 15 Terminate: Some process (or the batch system) has told this process to end
[1]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[1]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[1]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[1]PETSC ERROR: likely location of problem given in stack below
[1]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[1]PETSC ERROR: Note: The EXACT line numbers in the stack are not available,
[1]PETSC ERROR:       INSTEAD the line number of the start of the function
[1]PETSC ERROR:       is given.
[1]PETSC ERROR: [1] PetscSleep line 29 /home/fan/softwares/petsc/src/sys/utils/psleep.c
[1]PETSC ERROR: [1] PetscTraceBackErrorHandler line 178 /home/fan/softwares/petsc/src/sys/error/errtrace.c
[1]PETSC ERROR: [1] PetscError line 340 /home/fan/softwares/petsc/src/sys/error/err.c
[1]PETSC ERROR: [1] MatGetFactor line 4404 /home/fan/softwares/petsc/src/mat/interface/matrix.c
[1]PETSC ERROR: [1] PCSetUp_ILU line 82 /home/fan/softwares/petsc/src/ksp/pc/impls/factor/ilu/ilu.c
[1]PETSC ERROR: [1] PCSetUp line 856 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[1]PETSC ERROR: [1] KSPSetUp line 289 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[1]PETSC ERROR: [1] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[1]PETSC ERROR: [1] __mfem_pc_shell_apply line 4013 /home/fan/Nutstore Files/github_nutstore/mfem_source_code/linalg/petsc.cpp
[1]PETSC ERROR: [1] PCSHELL user function apply() line 115 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[1]PETSC ERROR: [1] PCApply_Shell line 112 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[1]PETSC ERROR: [1] PCApply line 426 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[1]PETSC ERROR: [1] KSP_PCApply line 279 /home/fan/softwares/petsc/include/petsc/private/kspimpl.h
[1]PETSC ERROR: [1] KSPInitialResidual line 40 /home/fan/softwares/petsc/src/ksp/ksp/interface/itres.c
[1]PETSC ERROR: [1] KSPSolve_GMRES line 225 /home/fan/softwares/petsc/src/ksp/ksp/impls/gmres/gmres.c
[1]PETSC ERROR: [1] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[1]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[1]PETSC ERROR: Signal received
[1]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[1]PETSC ERROR: Petsc Release Version 3.12.3, unknown 
[1]PETSC ERROR: Unknown Name on a arch-linux2-c-debug named yoga by fan Sat Apr  4 09:19:39 2020
[1]PETSC ERROR: Configure options --download-fblaslapack=1
[1]PETSC ERROR: #1 User provided function() line 0 in  unknown file
[2]PETSC ERROR: ------------------------------------------------------------------------
[2]PETSC ERROR: Caught signal number 15 Terminate: Some process (or the batch system) has told this process to end
[2]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[2]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[2]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[2]PETSC ERROR: likely location of problem given in stack below
[2]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[2]PETSC ERROR: Note: The EXACT line numbers in the stack are not available,
[2]PETSC ERROR:       INSTEAD the line number of the start of the function
[2]PETSC ERROR:       is given.
[2]PETSC ERROR: [2] PetscSleep line 29 /home/fan/softwares/petsc/src/sys/utils/psleep.c
[2]PETSC ERROR: [2] PetscTraceBackErrorHandler line 178 /home/fan/softwares/petsc/src/sys/error/errtrace.c
[2]PETSC ERROR: [2] PetscError line 340 /home/fan/softwares/petsc/src/sys/error/err.c
[2]PETSC ERROR: [2] MatGetFactor line 4404 /home/fan/softwares/petsc/src/mat/interface/matrix.c
[2]PETSC ERROR: [2] PCSetUp_ILU line 82 /home/fan/softwares/petsc/src/ksp/pc/impls/factor/ilu/ilu.c
[2]PETSC ERROR: [2] PCSetUp line 856 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[2]PETSC ERROR: [2] KSPSetUp line 289 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[2]PETSC ERROR: [2] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[2]PETSC ERROR: [2] __mfem_pc_shell_apply line 4013 /home/fan/Nutstore Files/github_nutstore/mfem_source_code/linalg/petsc.cpp
[2]PETSC ERROR: [2] PCSHELL user function apply() line 115 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[2]PETSC ERROR: [2] PCApply_Shell line 112 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[2]PETSC ERROR: [2] PCApply line 426 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[2]PETSC ERROR: [2] KSP_PCApply line 279 /home/fan/softwares/petsc/include/petsc/private/kspimpl.h
[2]PETSC ERROR: [2] KSPInitialResidual line 40 /home/fan/softwares/petsc/src/ksp/ksp/interface/itres.c
[2]PETSC ERROR: [2] KSPSolve_GMRES line 225 /home/fan/softwares/petsc/src/ksp/ksp/impls/gmres/gmres.c
[2]PETSC ERROR: [2] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[2]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[2]PETSC ERROR: Signal received
[2]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[2]PETSC ERROR: Petsc Release Version 3.12.3, unknown 
[2]PETSC ERROR: Unknown Name on a arch-linux2-c-debug named yoga by fan Sat Apr  4 09:19:39 2020
[2]PETSC ERROR: Configure options --download-fblaslapack=1
[2]PETSC ERROR: #1 User provided function() line 0 in  unknown file
[3]PETSC ERROR: ------------------------------------------------------------------------
[3]PETSC ERROR: Caught signal number 15 Terminate: Some process (or the batch system) has told this process to end
[3]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[3]PETSC ERROR: or see https://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[3]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[3]PETSC ERROR: likely location of problem given in stack below
[3]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[3]PETSC ERROR: Note: The EXACT line numbers in the stack are not available,
[3]PETSC ERROR:       INSTEAD the line number of the start of the function
[3]PETSC ERROR:       is given.
[3]PETSC ERROR: [3] PetscSleep line 29 /home/fan/softwares/petsc/src/sys/utils/psleep.c
[3]PETSC ERROR: [3] PetscTraceBackErrorHandler line 178 /home/fan/softwares/petsc/src/sys/error/errtrace.c
[3]PETSC ERROR: [3] PetscError line 340 /home/fan/softwares/petsc/src/sys/error/err.c
[3]PETSC ERROR: [3] MatGetFactor line 4404 /home/fan/softwares/petsc/src/mat/interface/matrix.c
[3]PETSC ERROR: [3] PCSetUp_ILU line 82 /home/fan/softwares/petsc/src/ksp/pc/impls/factor/ilu/ilu.c
[3]PETSC ERROR: [3] PCSetUp line 856 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[3]PETSC ERROR: [3] KSPSetUp line 289 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[3]PETSC ERROR: [3] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[3]PETSC ERROR: [3] __mfem_pc_shell_apply line 4013 /home/fan/Nutstore Files/github_nutstore/mfem_source_code/linalg/petsc.cpp
[3]PETSC ERROR: [3] PCSHELL user function apply() line 115 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[3]PETSC ERROR: [3] PCApply_Shell line 112 /home/fan/softwares/petsc/src/ksp/pc/impls/shell/shellpc.c
[3]PETSC ERROR: [3] PCApply line 426 /home/fan/softwares/petsc/src/ksp/pc/interface/precon.c
[3]PETSC ERROR: [3] KSP_PCApply line 279 /home/fan/softwares/petsc/include/petsc/private/kspimpl.h
[3]PETSC ERROR: [3] KSPInitialResidual line 40 /home/fan/softwares/petsc/src/ksp/ksp/interface/itres.c
[3]PETSC ERROR: [3] KSPSolve_GMRES line 225 /home/fan/softwares/petsc/src/ksp/ksp/impls/gmres/gmres.c
[3]PETSC ERROR: [3] KSPSolve line 656 /home/fan/softwares/petsc/src/ksp/ksp/interface/itfunc.c
[3]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[3]PETSC ERROR: Signal received
[3]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[3]PETSC ERROR: Petsc Release Version 3.12.3, unknown 
[3]PETSC ERROR: Unknown Name on a arch-linux2-c-debug named yoga by fan Sat Apr  4 09:19:39 2020
[3]PETSC ERROR: Configure options --download-fblaslapack=1
[3]PETSC ERROR: #1 User provided function() line 0 in  unknown file
[yoga:11806] 3 more processes have sent help message help-mpi-api.txt / mpi-abort
[yoga:11806] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages

```

**Comment by stefanozampini** [comment]
@fanronghong I will also need the `petsc_example_opts` file to run your code.
Have you tried running on valgrind?

**Comment by fanronghong** [comment]
Thank you for your reply @stefanozampini ,

Sorry about forgetting `petsc_example_opts`.
[petsc_example_opts.txt](https://github.com/mfem/mfem/files/4436557/petsc_example_opts.txt)

I didn't use valgrind, but I tried to debug step by step. It's something wrong with executing 
deconstructor of `PetscParMatrix B` after     
```
MFEMFinalizePetsc();
MPI_Finalize();
```

**Comment by fanronghong** [comment]
I just tried with valgrind and got below  message which seems something wrong in `MPI_Init(&argc, &argv);`
```
*** The MPI_Comm_rank() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[yoga:3603] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
==3603== 
==3603== HEAP SUMMARY:
==3603==     in use at exit: 17,219,930 bytes in 102,967 blocks
==3603==   total heap usage: 216,114 allocs, 113,147 frees, 42,780,543 bytes allocated
==3603== 
==3603== 8 bytes in 1 blocks are definitely lost in loss record 63 of 851
==3603==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC317346: ???
==3603==    by 0xC33BCA9: ???
==3603==    by 0xC30FCB4: ???
==3603==    by 0xBCFD473: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 10 bytes in 1 blocks are definitely lost in loss record 68 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC325DAD: ???
==3603==    by 0xC3244FD: ???
==3603==    by 0xC33A2EB: ???
==3603==    by 0xC339004: ???
==3603==    by 0x8A41208: opal_libevent2022_event_base_loop (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0xC336FCC: ???
==3603==    by 0x7A236DA: start_thread (pthread_create.c:463)
==3603==    by 0x848388E: clone (clone.S:95)
==3603== 
==3603== 10 bytes in 1 blocks are definitely lost in loss record 69 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x83FF9B9: strdup (strdup.c:42)
==3603==    by 0x8A26AA8: mca_base_var_enum_create_flag (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A3A6AF: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29004: mca_base_framework_register (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29080: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A290F0: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x509D4A2: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 10 bytes in 1 blocks are definitely lost in loss record 70 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x83FF9B9: strdup (strdup.c:42)
==3603==    by 0x13F89ED0: ???
==3603==    by 0x4010732: call_init (dl-init.c:72)
==3603==    by 0x4010732: _dl_init (dl-init.c:119)
==3603==    by 0x40151FE: dl_open_worker (dl-open.c:522)
==3603==    by 0x84C92DE: _dl_catch_exception (dl-error-skeleton.c:196)
==3603==    by 0x40147C9: _dl_open (dl-open.c:605)
==3603==    by 0x940AF95: dlopen_doit (dlopen.c:66)
==3603==    by 0x84C92DE: _dl_catch_exception (dl-error-skeleton.c:196)
==3603==    by 0x84C936E: _dl_catch_error (dl-error-skeleton.c:215)
==3603==    by 0x940B734: _dlerror_run (dlerror.c:162)
==3603==    by 0x940B050: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==3603== 
==3603== 17 bytes in 1 blocks are definitely lost in loss record 136 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x83FF9B9: strdup (strdup.c:42)
==3603==    by 0x8A26AA8: mca_base_var_enum_create_flag (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A3A6C9: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29004: mca_base_framework_register (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29080: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A290F0: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x509D4A2: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 17 bytes in 1 blocks are definitely lost in loss record 137 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x83FF9B9: strdup (strdup.c:42)
==3603==    by 0x8A26AA8: mca_base_var_enum_create_flag (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x509D876: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 35 bytes in 1 blocks are definitely lost in loss record 266 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x849648F: __vasprintf_chk (vasprintf_chk.c:80)
==3603==    by 0x849632E: __asprintf_chk (asprintf_chk.c:32)
==3603==    by 0x509D671: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 40 bytes in 1 blocks are definitely lost in loss record 289 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xE07F81A: ???
==3603==    by 0xE07EAA8: ???
==3603==    by 0xE006779: ???
==3603==    by 0xDFE4D0A: ???
==3603==    by 0xDDAAB0D: ???
==3603==    by 0x8EAF4C2: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8EA6910: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8E9F020: hwloc_topology_set_xmlbuffer (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0xBCFD91D: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603== 
==3603== 40 bytes in 1 blocks are definitely lost in loss record 290 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xE07F81A: ???
==3603==    by 0xE023169: ???
==3603==    by 0xE00679A: ???
==3603==    by 0xDFE4D0A: ???
==3603==    by 0xDDAAB0D: ???
==3603==    by 0x8EAF4C2: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8EA6910: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8E9F020: hwloc_topology_set_xmlbuffer (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0xBCFD91D: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603== 
==3603== 48 bytes in 1 blocks are definitely lost in loss record 333 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x8A1D9CC: mca_base_component_repository_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A1CA09: mca_base_component_find (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A28BC9: mca_base_framework_components_register (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29019: mca_base_framework_register (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A29080: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A02B7A: opal_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8767078: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 56 bytes in 1 blocks are definitely lost in loss record 356 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x11127037: ???
==3603==    by 0x11330884: ???
==3603==    by 0x10F0C128: ???
==3603==    by 0x87A352C: orte_oob_base_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x8794C04: orte_ess_base_app_setup (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0xBCFDE88: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 56 bytes in 1 blocks are definitely lost in loss record 357 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x5079EF7: ompi_attr_set_fortran_mpi1 (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x507BC5E: ompi_attr_create_predefined (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x509D88D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 56 bytes in 1 blocks are definitely lost in loss record 358 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x5079EF7: ompi_attr_set_fortran_mpi1 (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50AAEF8: PMPI_Add_error_class (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x549108B: PetscInitialize (pinit.c:910)
==3603==    by 0x2B1904: mfem::MFEMInitializePetsc(int*, char***, char const*, char const*) (petsc.cpp:158)
==3603==    by 0x24E3C4: main (petsc_example.cpp:88)
==3603== 
==3603== 68 bytes in 12 blocks are definitely lost in loss record 369 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x83FF9B9: strdup (strdup.c:42)
==3603==    by 0x13878087: ???
==3603==    by 0x13878822: ???
==3603==    by 0x138790E4: ???
==3603==    by 0x136674E7: ???
==3603==    by 0x50F74AF: ompi_mtl_base_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x132559C9: ???
==3603==    by 0x50FF9A4: mca_pml_base_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x509D544: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 72 bytes in 1 blocks are definitely lost in loss record 384 of 851
==3603==    at 0x4C31D2F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x849644E: __vasprintf_chk (vasprintf_chk.c:88)
==3603==    by 0x849632E: __asprintf_chk (asprintf_chk.c:32)
==3603==    by 0xBCFE368: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 88 bytes in 1 blocks are definitely lost in loss record 391 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xA0544A8: lt__malloc (in /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.1)
==3603==    by 0xA0544DD: lt__zalloc (in /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.1)
==3603==    by 0xA056779: ??? (in /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.1)
==3603==    by 0xA057120: lt_dlopenadvise (in /usr/lib/x86_64-linux-gnu/libltdl.so.7.3.1)
==3603==    by 0xB4E8E85: ???
==3603==    by 0x8EA5AE0: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8E9EE86: hwloc_topology_init (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0xBCFD8FA: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603== 
==3603== 104 bytes in 1 blocks are definitely lost in loss record 404 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xE07F8CA: ???
==3603==    by 0xE0E9BBC: ???
==3603==    by 0x7A2B826: __pthread_once_slow (pthread_once.c:116)
==3603==    by 0xE07FC70: ???
==3603==    by 0xE07F1B8: ???
==3603==    by 0xE00677E: ???
==3603==    by 0xDFE4D0A: ???
==3603==    by 0xDDAAB0D: ???
==3603==    by 0x8EAF4C2: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8EA6910: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8E9F020: hwloc_topology_set_xmlbuffer (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603== 
==3603== 371 (56 direct, 315 indirect) bytes in 1 blocks are definitely lost in loss record 451 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC34C6F5: ???
==3603==    by 0xC31014D: ???
==3603==    by 0x8794CC2: orte_ess_base_app_setup (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0xBCFDE88: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 447 (56 direct, 391 indirect) bytes in 1 blocks are definitely lost in loss record 452 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xB8F3731: ??? (in /tmp/openmpi-sessions-1000@yoga_0/12804/pmix_dstor_3604/smseg-839122945-0)
==3603==    by 0x8A1E129: mca_base_framework_components_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A290F0: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8767171: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 448 (232 direct, 216 indirect) bytes in 1 blocks are definitely lost in loss record 453 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x125FF02F: ???
==3603==    by 0x12E14C19: ???
==3603==    by 0x509D8F4: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 625 (56 direct, 569 indirect) bytes in 1 blocks are definitely lost in loss record 458 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC33B820: ???
==3603==    by 0x8A41208: opal_libevent2022_event_base_loop (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0xC336FCC: ???
==3603==    by 0x7A236DA: start_thread (pthread_create.c:463)
==3603==    by 0x848388E: clone (clone.S:95)
==3603== 
==3603== 726 (720 direct, 6 indirect) bytes in 1 blocks are definitely lost in loss record 461 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x11BE12B4: ???
==3603==    by 0x87A1C7C: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x87A2E27: orte_oob_base_set_addr (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x8A41208: opal_libevent2022_event_base_loop (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A0617D: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x7A236DA: start_thread (pthread_create.c:463)
==3603==    by 0x848388E: clone (clone.S:95)
==3603== 
==3603== 910 (400 direct, 510 indirect) bytes in 1 blocks are definitely lost in loss record 470 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xDFDF47F: ???
==3603==    by 0xE0067A4: ???
==3603==    by 0xDFE4D0A: ???
==3603==    by 0xDDAAB0D: ???
==3603==    by 0x8EAF4C2: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8EA6910: ??? (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0x8E9F020: hwloc_topology_set_xmlbuffer (in /usr/lib/x86_64-linux-gnu/libhwloc.so.5.7.6)
==3603==    by 0xBCFD91D: ???
==3603==    by 0x87671DD: orte_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-rte.so.20.10.1)
==3603==    by 0x509D27D: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603== 
==3603== 1,120 bytes in 1 blocks are definitely lost in loss record 475 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x149AB6B6: ???
==3603==    by 0x50E2172: mca_coll_base_comm_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x509DC48: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 1,120 bytes in 1 blocks are definitely lost in loss record 476 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x149AB6B6: ???
==3603==    by 0x50E2172: mca_coll_base_comm_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x509DC67: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== 1,120 bytes in 1 blocks are definitely lost in loss record 477 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x149AB6B6: ???
==3603==    by 0x50E2172: mca_coll_base_comm_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50815AD: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x5083F24: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x8A019EB: opal_progress (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x50839B4: ompi_comm_activate (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x507E81D: ompi_comm_dup_with_info (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50B25EE: PMPI_Comm_dup (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x5457CC9: PetscCommDuplicate (tagm.c:126)
==3603==    by 0x545E734: PetscHeaderCreate_Private (inherit.c:64)
==3603==    by 0x59F9B1D: VecCreate (veccreate.c:34)
==3603== 
==3603== 1,120 bytes in 1 blocks are definitely lost in loss record 478 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x149AB6B6: ???
==3603==    by 0x50E2172: mca_coll_base_comm_select (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50815AD: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x5083F24: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x8A019EB: opal_progress (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x50839B4: ompi_comm_activate (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x507E81D: ompi_comm_dup_with_info (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50B25EE: PMPI_Comm_dup (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x5457CC9: PetscCommDuplicate (tagm.c:126)
==3603==    by 0x545E734: PetscHeaderCreate_Private (inherit.c:64)
==3603==    by 0x5974DF9: ISCreate (isreg.c:37)
==3603== 
==3603== 4,152 bytes in 1 blocks are definitely lost in loss record 704 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC359303: ???
==3603==    by 0xC35AD3C: ???
==3603==    by 0xC33FBBA: ???
==3603==    by 0x8A41208: opal_libevent2022_event_base_loop (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0xC336FCC: ???
==3603==    by 0x7A236DA: start_thread (pthread_create.c:463)
==3603==    by 0x848388E: clone (clone.S:95)
==3603== 
==3603== 4,152 bytes in 1 blocks are definitely lost in loss record 705 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0xC3594B6: ???
==3603==    by 0xC35AD3C: ???
==3603==    by 0xC33FBBA: ???
==3603==    by 0x8A41208: opal_libevent2022_event_base_loop (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0xC336FCC: ???
==3603==    by 0x7A236DA: start_thread (pthread_create.c:463)
==3603==    by 0x848388E: clone (clone.S:95)
==3603== 
==3603== 16,568 bytes in 1 blocks are definitely lost in loss record 772 of 851
==3603==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3603==    by 0x89F96AA: opal_free_list_grow_st (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x89FF9CE: opal_rb_tree_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x123F99AF: ???
==3603==    by 0x123FA1EC: ???
==3603==    by 0x8A1E129: mca_base_framework_components_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A551E1: ??? (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x8A290F0: mca_base_framework_open (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libopen-pal.so.20.10.1)
==3603==    by 0x509D482: ompi_mpi_init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x50BE2AA: PMPI_Init (in /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so.20.10.1)
==3603==    by 0x24E36C: main (petsc_example.cpp:83)
==3603== 
==3603== LEAK SUMMARY:
==3603==    definitely lost: 31,607 bytes in 41 blocks
==3603==    indirectly lost: 2,007 bytes in 60 blocks
==3603==      possibly lost: 0 bytes in 0 blocks
==3603==    still reachable: 17,186,316 bytes in 102,866 blocks
==3603==         suppressed: 0 bytes in 0 blocks
==3603== Reachable blocks (those to which a pointer was found) are not shown.
==3603== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==3603== 
==3603== For counts of detected and suppressed errors, rerun with: -v
==3603== Use --track-origins=yes to see where uninitialised values come from
==3603== ERROR SUMMARY: 31 errors from 31 contexts (suppressed: 0 from 0)
```

**Comment by stefanozampini** [comment]
This is definitely a different error than the one you previously posted.
The issue here is simple. The destructor of the PETSc objects gets called after MPI is finalized, so the error, since PETSc tries to free the communicator after MPI_Finalize has been called.
You have two alternatives. Either use

```
MFEMInitializePetsc(NULL, NULL, options_src, NULL);

{

PetscParVector x;

}// destructor of x gets called

MFEMFinalizePetsc()
```

or
```
MFEMInitializePetsc(NULL, NULL, options_src, NULL);

PetscParVector *x = new PetscParVector();

....// do something with x

delete x; // x destructor gets called

MFEMFinalizePetsc()
```

**Comment by fanronghong** [comment]
Sorry about the wrong error @stefanozampini ,

For the 1st solution, you mean I should put all "petsc objects (not pointers)" (like `PetscParVector`, `PetscParMatrix`, `PetscLinearSolver` ...) between `MFEMInitializePetsc(NULL, NULL, options_src, NULL);` and `MFEMFinalizePetsc();` . Am I right?

For the 2nd solution, I should put all "petsc pointers (not objects)" between `MFEMInitializePetsc(NULL, NULL, options_src, NULL);` and `delete pointers; MFEMFinalizePetsc();`. Am I right?

**Comment by stefanozampini** [comment]
In C++, statically allocated objects get destroyed when they go out of scope (see e.g. https://stackoverflow.com/questions/10081429/when-is-a-c-destructor-called). This is why you need to instantiate PETSc-based objects within curly braces if within the `main` method (1st solution). Instead, if you allocate them dynamically (2nd solution), you have to make sure you call `delete` on them before `MFEMFinalizePetsc()`

**Comment by fanronghong** [comment]
Thank you very much @stefanozampini , your answer is a great help.
